saleor/graphql/product/mutations/products.py

```diff
@@ -846,28 +846,35 @@ def clean_input(
                 cleaned_input["product"]
             )
 
-        # Run the validation only if product type is configurable
-        if product_type.has_variants:
-            # Attributes are provided as list of `AttributeValueInput` objects.
-            # We need to transform them into the format they're stored in the
-            # `Product` model, which is HStore field that maps attribute's PK to
-            # the value's PK.
-            attributes = cleaned_input.get("attributes")
-            try:
-                if attributes:
-                    cleaned_attributes = cls.clean_attributes(attributes, product_type)
-                    cls.validate_duplicated_attribute_values(
-                        cleaned_attributes, used_attribute_values, instance
-                    )
-                    cleaned_input["attributes"] = cleaned_attributes
-                elif not instance.pk and not attributes:
-                    # if attributes were not provided on creation
-                    raise ValidationError(
-                        "All attributes must take a value.",
-                        ProductErrorCode.REQUIRED.value,
+        if not product_type.has_variants:
+            raise ValidationError(
+                {
+                    "product": ValidationError(
+                        "Only one variant can be assigned to this product.",
+                        code=ProductErrorCode.INVALID.value,
                     )
-            except ValidationError as exc:
-                raise ValidationError({"attributes": exc})
+                }
+            )
+        # Attributes are provided as list of `AttributeValueInput` objects.
+        # We need to transform them into the format they're stored in the
+        # `Product` model, which is HStore field that maps attribute's PK to
+        # the value's PK.
+        attributes = cleaned_input.get("attributes")
+        try:
+            if attributes:
+                cleaned_attributes = cls.clean_attributes(attributes, product_type)
+                cls.validate_duplicated_attribute_values(
+                    cleaned_attributes, used_attribute_values, instance
+                )
+                cleaned_input["attributes"] = cleaned_attributes
+            elif not instance.pk and not attributes:
+                # if attributes were not provided on creation
+                raise ValidationError(
+                    "All attributes must take a value.",
+                    ProductErrorCode.REQUIRED.value,
+                )
+        except ValidationError as exc:
+            raise ValidationError({"attributes": exc})
```

Comment:
I see that `clean_input` is quite long. Maybe we can separate this to another function.